oak_client/src/lib.rs

```diff
@@ -22,34 +22,92 @@ pub mod proto {
 pub mod transport;
 pub mod verifier;
 
-use anyhow::Context;
+use crate::transport::AsyncEvidenceProvider;
+use anyhow::{anyhow, Context};
+use oak_crypto::{
+    schema::{AeadEncryptedMessage, HpkeRequest, HpkeResponse},
+    SenderCryptoProvider,
+};
+use prost::Message;
 use std::vec::Vec;
 
+const EMPTY_ASSOCIATED_DATA: &[u8] = b"";
+
 /// Client for connecting to Oak.
 /// Represents a Relying Party from the RATS Architecture:
 /// <https://www.rfc-editor.org/rfc/rfc9334.html#name-relying-party>
 pub struct OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
     transport: T,
+    crypto_provider: SenderCryptoProvider,
 }
 
 impl<T> OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
-    pub async fn create(transport: T) -> anyhow::Result<Self> {
+    pub async fn create(mut transport: T) -> anyhow::Result<Self> {
         // TODO(#3442): Fetch enclave public key.
         // TODO(#3641): Implement client-side attestation verification.
-        Ok(Self { transport })
+        let evidence = transport
+            .get_evidence()
```

Comment:
It seems overkill to pass an entire AsyncEvidenceProvider -- why not just pass the evidence, obtained by the caller? Alongside the transport (which would not require the extra trait bound at all).

Reply:
This would mean that we'll need to create a second `EvidenceRequestor` client which only requests the evidence instead of hiding all of the logic inside `OakClient`. And this `EvidenceRequestor` client would also consume the gRPC transport.

Reply:
Is it not possible to invoke it without consuming it? I thought a mut reference would be enough.

Reply:
This would still mean that we won't have encapsulation of the client logic in a single object.